### PR TITLE
Improve `InternedString` usability

### PR DESCRIPTION
**Pull request by oli-obk**

### CODE REVIEW

File: src/libcollections/linked_list.rs [hunk 1]
```diff
@@ -1376,7 +1376,7 @@ mod tests {
         thread::spawn(move || {
                 check_links(&n);
                 let a: &[_] = &[&1, &2, &3];
-                assert_eq!(a, &n.iter().collect::<Vec<_>>()[..]);
+                assert_eq!(a, &n.iter().collect::<Vec<_>>());
```
Comment (jseyfried) [inline src/libcollections/linked_list.rs]: This is causing Travis to fail -- the second argument isn't getting autoderefed.

**Review by jseyfried — COMMENTED** [review]
r=me modulo Travis failure

**Comment by oli-obk** [comment]
travis likes this now

**Comment by jseyfried** [comment]
@oli-obk needs another rebase -- ping me when ready.

**Comment by oli-obk** [comment]
ping @jseyfried just a conflict with the now removed liblog

**Comment by oli-obk** [comment]
The compiler panicked while running some libstd tests on musl.

```
[01:17:04] error: process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-std/i686-unknown-linux-musl/release/deps/std-d02fbe97fc497570` (signal: 11, SIGSEGV: invalid memory reference)
[01:17:04] 
[01:17:04] To learn more, run the command again with --verbose.
[01:17:04] 
[01:17:04] 
[01:17:04] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "-j" "4" "--target" "i686-unknown-linux-musl" "--release" "--locked" "--manifest-path" "/checkout/src/libstd/Cargo.toml" "--features" "panic-unwind jemalloc backtrace" "-p" "std:0.0.0" "-p" "rustc_msan:0.0.0" "-p" "alloc_system:0.0.0" "-p" "core:0.0.0" "-p" "unwind:0.0.0" "-p" "rustc_asan:0.0.0" "-p" "rustc_lsan:0.0.0" "-p" "collections:0.0.0" "-p" "rustc_tsan:0.0.0" "-p" "alloc:0.0.0" "-p" "libc:0.0.0" "-p" "rand:0.0.0" "-p" "compiler_builtins:0.0.0" "-p" "panic_abort:0.0.0" "-p" "std_unicode:0.0.0" "--"
[01:17:04] expected success, got: exit code: 101
```

spurious? Or some pre-existing bug? I didn't touch any unsafe code.

**Comment by alexcrichton** [comment]
Unfortunately a spurious error -- https://github.com/rust-lang/rust/issues/38618

@bors: retry

**Comment by zackmdavis** [comment]
:heart:

**Pull request merged**